tests/test_registries.py

```diff
@@ -0,0 +1,166 @@
+#
+# (c) Copyright 2017-2018 Hewlett Packard Enterprise Development LP
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+
+import unittest
+
+import fixtures
+import testtools
+
+from . import base  # noqa: H304
+
+import windlass.registries
+
+
+class TestRegistries(testtools.TestCase):
+
+    def test_from_url_aws_ecr(self):
+        registry_url = "000000000000.dkr.ecr.us-west-1.amazonaws.com"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(registry, windlass.registries.EcrDockerRegistry)
+
+    def test_from_url_jfrog(self):
+        registry_url = "example-org.jfrog.io"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(
+            registry, windlass.registries.JfrogDockerRegistry
+        )
+
+    def test_from_url_hub(self):
+        registry_url = "registry.hub.docker.com"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(registry, windlass.registries.DockerRegistry)
+
+    def test_from_url_other(self):
+        registry_url = "registry.quay.io"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(registry, windlass.registries.DockerRegistry)
+
+
+class TestDockerRegistry(testtools.TestCase):
+
+    reg_url = "registry.hub.docker.com"
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_USER', 'my_user')
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_PASSWORD',
+                     'my_password')
+    def test_credentials_from_env(self):
+        registry = windlass.registries.DockerRegistry(self.reg_url)
+
+        self.assertEqual('my_user', registry.username)
+        self.assertEqual('my_password', registry.password)
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_USER', 'my_user')
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_PASSWORD',
+                     'my_password')
+    def test_credentials_ignore_env_when_set(self):
+        registry = windlass.registries.DockerRegistry(
+            self.reg_url,
+            "explicit_user",
+            "explicit_password",
+        )
+
+        self.assertEqual('explicit_user', registry.username)
+        self.assertEqual('explicit_password', registry.password)
+
+
+@unittest.mock.patch('windlass.registries.remotes.ECRConnector')
```

Comment:
I would also recommend to drop this and use fixture for mocking. I did this is in many tests in publishing as it really simplifies reading and if we are already agreed on using fixtures makes test code much more consistent.

Reply:
Done